src/auto-evo/simulation/PopulationSimulation.cs

```diff
@@ -117,36 +117,132 @@ private static void RunSimulationStep(SimulationConfiguration parameters, List<S
-            // when there are many species
-            bool lowSpecies = species.Count <= Constants.AUTO_EVO_LOW_SPECIES_THRESHOLD;
-            bool highSpecies = species.Count >= Constants.AUTO_EVO_HIGH_SPECIES_THRESHOLD;
+            var speciesEnergies = new Dictionary<MicrobeSpecies, float>(species.Count);
+
+            var totalPhotosynthesisScore = 0.0f;
+            var totalChemosynthesisScore = 0.0f;
+            var totalPredationScore = 0.0f;
+
+            // Calculate the total scores of each type in the current patch
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                totalPhotosynthesisScore += GetPhotosynthesisScore(currentMicrobeSpecies);
+                totalChemosynthesisScore += GetChemosynthesisScore(currentMicrobeSpecies);
+                totalPredationScore += GetPredationScore(currentMicrobeSpecies);
+            }
+
+            // Avoid division by 0
+            totalPhotosynthesisScore = Math.Max(MathUtils.EPSILON, totalPhotosynthesisScore);
+            totalChemosynthesisScore = Math.Max(MathUtils.EPSILON, totalChemosynthesisScore);
+            totalPredationScore = Math.Max(MathUtils.EPSILON, totalPredationScore);
+
+            // Calculate the share of environmental energy captured by each species
+            var energyAvailableForPredation = 0.0f;
+
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                var currentSpeciesEnergy = 0.0f;
+
+                currentSpeciesEnergy += sunlightInPatch
+                    * GetPhotosynthesisScore(currentMicrobeSpecies) / totalPhotosynthesisScore;
+
+                currentSpeciesEnergy += hydrogenSulfideInPatch
+                    * GetChemosynthesisScore(currentMicrobeSpecies) / totalChemosynthesisScore;
+
+                energyAvailableForPredation += 0.5f * currentSpeciesEnergy;
+                speciesEnergies.Add(currentMicrobeSpecies, currentSpeciesEnergy);
+            }
+
+            // Calculate the share of predation done by each species
+            // Then update populations
+            foreach (MicrobeSpecies currentMicrobeSpecies in species)
+            {
+                speciesEnergies[currentMicrobeSpecies] += energyAvailableForPredation
+                    * GetPredationScore(currentMicrobeSpecies) / totalPredationScore;
+
+                var newPopulation = (int)(speciesEnergies[currentMicrobeSpecies]
+                    / Math.Pow(currentMicrobeSpecies.Organelles.Count(), 1.3f));
+
+                populations.AddPopulationResultForSpecies(currentMicrobeSpecies, patch, newPopulation);
+            }
+        }
+
+        private static float GetPhotosynthesisScore(MicrobeSpecies species)
+        {
+            var photosynthesisScore = 0.0f;
+
+            var sunlight = SimulationParameters.Instance.GetCompound("sunlight");
+            var glucose = SimulationParameters.Instance.GetCompound("glucose");
```

Comment:
Constantly looking these up should be avoided, making these `private static readonly` fields should maybe work?
I think those get initialized on the first read.